### PR TITLE
Add codes to generate the new url to return Grinder in fail tests.

### LINKED ISSUE #3219: Create Rerun failed tests in Grinder link
**Background**: 
We have  `Rerun in Grinder` link in Jenkins job. It captures all parameters and allows the users to rerun the job easily.

image

https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/

However, one common use case is that users need to rerun for failed test target(s). `TARGET` parameter captures the test target for the build. If the users need to run the specific failed target(s), they need to modify the `TARGET` value to the specific failed target(s). This takes time and is not very convenient for the users. I think we should further simplify the workflow for the users.

**Proposal**:
Create `Rerun failed tests in Grinder` link. Instead of generating the `Rerun in Grinder` link with `TARGET` matches with the test build, we should generate `TARGET` value for the failed test(s). 

**Details**:
After the test execution, we can get failed test target from one of two places (failed.mk or *.tap file). We can have a parser to get the failed test target name(s). Set the` TARGET` value in the `Rerun failed tests in Grinder` link. For simplification, we should use testlist regardless of the number of failed tests. For example:

```
TARGET=testList TESTLIST=jit_jitt,jit_recognizedMethod_1,abcTest_2
```

### PR DESCRIPTION
**Pull request by YeeSkywalker**
Date: Feb 16th
Author: Yee Lin
Issue: [#3219](https://github.com/adoptium/aqa-tests/issues/3219)
Closes #3219

### CODE REVIEW

File: buildenv/jenkins/JenkinsfileBase
Comment (smlambert) [inline buildenv/jenkins/JenkinsfileBase:76]: hi @YeeSkywalker - please change "FAIL_TEST" as the name, since it sounds like a directive.  What this environment variable is really representing/storing is the "FAILED_TESTLIST".

**Review by smlambert — COMMENTED** [review]
thanks @YeeSkywalker - I have added some notes (before you have likely even run some Grinders to test the behaviour.  Feel free to move your PR to Draft while you work on updates and testing of it.

**Comment by llxia** [comment]
Converted this PR to draft as this is still WIP.

**Comment by YeeSkywalker** [comment]
I ran the Grinder test [#3674](https://ci.adoptopenjdk.net/job/Grinder/3674/) and it failed due to lack of permission. The output console indicated that `org.tap4j.consumer.TapConsumerFactory` was not usable in this Jenkins Groovy sandbox. It seemed like admins might need to grant permission for this plugin. I have tested `org.tap4j.consumer.TapConsumerFactory`, `org.tap4j.consumer.TapConsumer` and `org.tap4j.model.TestSet` locally, and they worked as expect. I wonder if I can solve this exception on my own or wait for approval of the tap4j plugin?

**Comment by smlambert** [comment]
re: https://github.com/adoptium/aqa-tests/pull/3362#issuecomment-1047653842 after script approval, rerun in [Grinder 3687](https://ci.adoptopenjdk.net/job/Grinder/3687)

**Comment by YeeSkywalker** [comment]
Thanks @smlambert! Unfortunately I was not able to rerun Grinder 3687. When I clicked the link it just returned a 404 page.

**Comment by YeeSkywalker** [comment]
> re: [#3362 (comment)](https://github.com/adoptium/aqa-tests/pull/3362#issuecomment-1047653842) after script approval, rerun in [Grinder 3687](https://ci.adoptopenjdk.net/job/Grinder/3687)
> 
> Other script approvals needed, rerun in [Grinder 3694](https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/3694)
Thanks @smlambert! Unfortunately I was not able to rerun Grinder 3687. When I clicked the link it just returned a 404 page.

**Comment by YeeSkywalker** [comment]
> re: [#3362 (comment)](https://github.com/adoptium/aqa-tests/pull/3362#issuecomment-1047653842) after script approval, rerun in [Grinder 3687](https://ci.adoptopenjdk.net/job/Grinder/3687)
> 
> Other script approvals needed, rerun in [Grinder 3694](https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/3694)

Thanks @smlambert! Unfortunately I was not able to rerun Grinder 3687. When I clicked the link it just returned a 404 page.

**Comment by YeeSkywalker** [comment]
The Grinder test link: [#3702](https://ci.adoptopenjdk.net/job/Grinder/3702/)
Since I don't have any failed test cases, I modified the code to find all pass tests and generated the pass test link. The console output is following the link: [3698](https://ci.adoptopenjdk.net/job/Grinder/3698/)

**Comment by YeeSkywalker** [comment]
Here are links of Grinder test with paralleled tests. [Grinder #3735](https://ci.adoptopenjdk.net/job/Grinder/3735/console) ran with `TARGET = sanity.openjdk`and returned result of 'UNSTABLE'.

**Comment by llxia** [comment]
re https://github.com/adoptium/aqa-tests/pull/3362#issuecomment-1054229069, could you point me at the generated `Rerun failed tests in Grinder` link?

**Comment by smlambert** [comment]
Neither child job of [Grinder 3735](https://ci.adoptopenjdk.net/job/Grinder/3735) seemed to have a Rerun failed tests link and I would have expected both children to have a link, since both contained a failing target (https://ci.adoptopenjdk.net/job/Grinder/3735/tapResults/).

**Comment by YeeSkywalker** [comment]
> 

After inserting tap file collecting helper function `Parallel Test` stage, the fail Grinder target will be `testList TESTLIST=jdk_jdi_jdk8_0,jdk_jdi_jdk8_1`. The link to the Grinder test is [Grinder 3746](https://ci.adoptopenjdk.net/job/Grinder/3746/)

**Comment by YeeSkywalker** [comment]
> After inserting tap file collecting helper function `Parallel Test` stage, the fail Grinder target will be `testList TESTLIST=jdk_jdi_jdk8_0,jdk_jdi_jdk8_1`. The link to the Grinder test is [Grinder 3746](https://ci.adoptopenjdk.net/job/Grinder/3746/)

After inserting tap file collecting helper function `Parallel Test` stage, the fail Grinder target will be `testList TESTLIST=jdk_jdi_jdk8_0,jdk_jdi_jdk8_1`. The link to the Grinder test is [Grinder 3746](https://ci.adoptopenjdk.net/job/Grinder/3746/)

**Comment by YeeSkywalker** [comment]
> Neither child job of [Grinder 3735](https://ci.adoptopenjdk.net/job/Grinder/3735) seemed to have a Rerun failed tests link and I would have expected both children to have a link, since both contained a failing target (https://ci.adoptopenjdk.net/job/Grinder/3735/tapResults/).
> 
> Child job links are [Grinder_testList_0/58](https://ci.adoptopenjdk.net/job/Grinder_testList_0/58/) and [Grinder_testList_1/35](https://ci.adoptopenjdk.net/job/Grinder_testList_1/35) respectively.

After inserting tap file collecting helper function `Parallel Test` stage, the fail Grinder target will be `testList TESTLIST=jdk_jdi_jdk8_0,jdk_jdi_jdk8_1`. The link to the Grinder test is [Grinder 3746](https://ci.adoptopenjdk.net/job/Grinder/3746/)

**Comment by smlambert** [comment]
Yes, I see the link in Grinder 3746 child jobs (https://ci.adoptopenjdk.net/job/Grinder_testList_1/40/ and https://ci.adoptopenjdk.net/job/Grinder_testList_0/63/).  The link reads "Return in Grinder with failed tests", will need to update to "Rerun in Grinder with failed test targets".  I would also like the variable names to read as nouns, not verbs so prefer FAILED_TESTLIST versus FAIL_TESTLIST.  I will put these review comments in line.  Its coming along well though!

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:773]: - please use a function, so we do not have to duplicate the code at line 1044
- please use a local variable for the failed test list (not environment variable) - `env.FAIL_TESTLIST` (please also rename as suggested by Shelley)

Comment (YeeSkywalker) [reply buildenv/jenkins/JenkinsfileBase:756]: After tidying the code, the reruning Grinder link is [Grinder 3776](https://ci.adoptopenjdk.net/job/Grinder/3776/)

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:756]: Could you add if block inside addFailedTestsGrinderLink()?

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:79]: Rename RERUN to RERUN_LINK

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:80]: Rename FAIL_TEST_TARGET to FAILED_TEST_TARGET

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:1013]: we do not have `env.FAIL_TESTLIST` anymore. Please update the echo message.

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:1014]: Please update the message as suggested: `Rerun in Grinder with failed test targets`, not `Return`

**Review by llxia — CHANGES_REQUESTED** [review]

**Comment by llxia** [comment]
image


For better user ability, please move `Grinder wiki ` before all Grinder links.
Move
https://github.com/adoptium/aqa-tests/blob/61bdf3f9bead4dd2369548a5c1dbada82dd5c0cd/buildenv/jenkins/JenkinsfileBase#L990
before line 985.

Please run Grinder again after all above changes. Thanks

**Comment by YeeSkywalker** [comment]
I have updated and tidied code as suggestions, and new parallel tests grinder link is [Grinder 3797](https://ci.adoptopenjdk.net/job/Grinder/3797/). Please review when you are free!

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:1009]: Please remove this echo

File: buildenv/jenkins/JenkinsfileBase
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:1013]: Please keep line 1012 and 1013 message to be the same - `Rerun in Grinder with failed test targets ` and the rerun  link (not just test list)

**Review by llxia — CHANGES_REQUESTED** [review]

**Comment by llxia** [comment]
https://ci.adoptopenjdk.net/job/Grinder/3808/

**Review by smlambert — APPROVED** [review]
Nicely done, thanks @YeeSkywalker !

**Comment by AdamBrousseau** [comment]
@llxia fyi I approved the following scripts
```
org.tap4j.consumer.TapConsumerFactory makeTap13Consumer
org.tap4j.consumer.TapConsumer load java.lang.String
```
on OpenJ9 and Internal (IBM) Jenkins.

**Comment by llxia** [comment]
Thanks @AdamBrousseau

**Pull request merged**
